File: 0016_3Sum_Closest/3Sum_Closest.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer}
def three_sum_closest(nums, target)
  return -1 if nums.length < 3
  result = 0
  distance = 1_073_741_823
  nums.sort!
  (0...nums.length - 2).each do |i|
    j = i + 1
    k = nums.length - 1
    while j < k
      temp = nums[i] + nums[j] + nums[k]
      if temp < target
        temp_dist = target - temp
        if temp_dist < distance
          result = nums[i] + nums[j] + nums[k]
          distance = temp_dist
        end
        j += 1
      elsif temp > target
        temp_dist = temp - target
        if temp_dist < distance
          result = nums[i] + nums[j] + nums[k]
          distance = temp_dist
        end
        k -= 1
      else
        return nums[i] + nums[j] + nums[k]
      end
    end
  end
  result
end

File: 0016_3Sum_Closest/3Sum_Closest.py
# -*- coding: utf-8 -*-

"""
Given an array nums of n integers and an integer target, find three integers in nums such that the sum is closest to target. Return the sum of the three integers. You may assume that each input would have exactly one solution.

Example:

Given array nums = [-1, 2, 1, -4], and target = 1.

The sum that is closest to the target is 2. (-1 + 2 + 1 = 2).
"""


class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        length = len(nums)
        flag = sys.maxsize
        result = 0
        for i in range(length - 2):
            left = i + 1
            right = length - 1
            while left < right:
                if target > (nums[i] + nums[left] + nums[right]):
                    if target - (nums[i] + nums[left] + nums[right]) < flag:
                        flag = target - (nums[i] + nums[left] + nums[right])
                        result = nums[i] + nums[left] + nums[right]
                    left += 1
                elif target < (nums[i] + nums[left] + nums[right]):
                    if (nums[i] + nums[left] + nums[right]) - target < flag:
                        flag = (nums[i] + nums[left] + nums[right]) - target
                        result = nums[i] + nums[left] + nums[right]
                    right -= 1
                else:
                    return target
        return result
